### ISSUE TITLE
If I want to run VGG16 based quantization, how do I run the code?

### ISSUE THREAD

**Issue by Jealousc11gx** [RuokaiYin/MINT_Quantization#1]
Hi! thanks for your work. I want to know that if I want to run this code in vgg, which params should I give ?

**Comment by RuokaiYin** [comment]
Hi! Thank you for your interest in our work. I plan to add some more README descriptions. I will do so in a couple of days...

I can first provide you with some running command lines. 
You may want to run the baseline model without quantization:
python3 snn_train.py --arch vgg16
(To make things more neat, please go to args_config.py to manually change the path of dataset_dir, dataset, learning_rate, etc..)

Then, to run the uniform quantized models 
python3 snn_train.py --arch vgg16 -wq -uq
the wq and uq flags turn on weight and membrane potential quantization.
If you are running VGG16, please go to the quant_net.py to modify the Q_ShareScale_VGG16 with the target bitwidth of num_bits_w and num_bits_u.

To run the MINT-quantized model:
python3 snn_train.py --arch vgg16 -wq -uq -share
the additional -share flag will turn on the quantization scaling share function, as we proposed in the paper.

I am sorry that the current code version is a bit messy. I will clean it up ASAP. In the meantime, please try the steps I commented on above, which should be enough to run the experiments. Please let me know if you have any further questions. Thank you!

**Comment by Jealousc11gx** [comment]
Thank a lot！I've already ran this code successfully!

**Comment by Jealousc11gx** [comment]
Another question is, can I quantize W and U to different number of bits?i.e w2u3? Have you ever done this related experiments?

**Comment by Jealousc11gx** [comment]
One more question... I failed to train the vgg9 when I set quantization bit to 8
<img width="1010" alt="d6a7ec516901a8e87f601f67922d9a8" src="https://github.com/RuokaiYin/MINT_Quantization/assets/67589956/31fc02c9-3f50-4eec-adfc-1272fa18259f">


root@autodl-container-bb8c4f8e12-9b6fb58b:~/autodl-tmp/quant_snn# python train_snn.py -uq -wq -share -sft_rst --arch vgg9
********** SNN simulation parameters **********
Namespace(batch_size=256, lr=0.001, arch='vgg9', dataset_dir='./dataset/', dataset='cifar10', optim='adam', leak_mem=0.5, th=0.5, rst='hard', T=4, uq=True, bq=False, wq=True, share=True, sft_rst=True, conv_b=False, bn_a=False, xa=False, ts=False, epoch=150, num_workers=4, train_display_freq=1, test_display_freq=1)
Files already downloaded and verified
Files already downloaded and verified
quant bw for w: 8
quant bw for u: 8
Train Epoch: 0/150 Loss: 2.303124 Accuracy: 10.000% Best Accuracy: 10.000%
Train Epoch: 1/150 Loss: 2.302700 Accuracy: 10.000% Best Accuracy: 10.000%
Train Epoch: 2/150 Loss: 2.302739 Accuracy: 10.000% Best Accuracy: 10.000%
Train Epoch: 3/150 Loss: 2.302704 Accuracy: 10.000% Best Accuracy: 10.000%
Train Epoch: 4/150 Loss: 2.302775 Accuracy: 10.000% Best Accuracy: 10.000%
Train Epoch: 5/150 Loss: 2.302424 Accuracy: 10.000% Best Accuracy: 10.000%
Train Epoch: 6/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000%
Train Epoch: 7/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000%
Train Epoch: 8/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000%
Train Epoch: 9/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000%

**Comment by ruokaiy-cerebras** [comment]
> Another question is, can I quantize W and U to different number of bits?i.e w2u3? Have you ever done this related experiments?

Yes. Algorithmically, quantizing W and U to different bitwidth makes sense. On the hardware level, having an odd number of bitwidth might not make that much sense, though. You must carefully tune your hardware design to fully leverage the odd number of bitwidth. I did not run the relevant experiments, and you can certainly try!

**Comment by RuokaiYin** [comment]
> Another question is, can I quantize W and U to different number of bits?i.e w2u3? Have you ever done this related experiments?

Yes. Algorithmically, quantizing W and U to different bitwidth makes sense. From the hardware perspective, having an odd number of bitwidth might not make much sense, though. You must carefully tune your hardware design to fully leverage the odd-number bitwidth. I did not run the relevant experiments, and you can certainly try!

**Comment by RuokaiYin** [comment]
> One more question... I failed to train the vgg9 when I set quantization bit to 8 <img alt="d6a7ec516901a8e87f601f67922d9a8" width="1010" src="https://private-user-images.githubusercontent.com/67589956/345525456-31fc02c9-3f50-4eec-adfc-1272fa18259f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwNjU4NDcsIm5iZiI6MTcyMDA2NTU0NywicGF0aCI6Ii82NzU4OTk1Ni8zNDU1MjU0NTYtMzFmYzAyYzktM2Y1MC00ZWVjLWFkZmMtMTI3MmZhMTgyNTlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA0VDAzNTkwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmMGE2ODhmYzk4MzU3MTkzOGNiMjg2M2VmMGEzM2I3NzMzYTgxMjMxNjg1NDUxN2JkYWQyZjM1Y2UyZTA0MDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uR5qliaX4bghW3DnabbceNQpOQfB8UPUJtqKRRL7k08">
> 
> root@autodl-container-bb8c4f8e12-9b6fb58b:~/autodl-tmp/quant_snn# python train_snn.py -uq -wq -share -sft_rst --arch vgg9 ********** SNN simulation parameters ********** Namespace(batch_size=256, lr=0.001, arch='vgg9', dataset_dir='./dataset/', dataset='cifar10', optim='adam', leak_mem=0.5, th=0.5, rst='hard', T=4, uq=True, bq=False, wq=True, share=True, sft_rst=True, conv_b=False, bn_a=False, xa=False, ts=False, epoch=150, num_workers=4, train_display_freq=1, test_display_freq=1) Files already downloaded and verified Files already downloaded and verified quant bw for w: 8 quant bw for u: 8 Train Epoch: 0/150 Loss: 2.303124 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 1/150 Loss: 2.302700 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 2/150 Loss: 2.302739 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 3/150 Loss: 2.302704 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 4/150 Loss: 2.302775 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 5/150 Loss: 2.302424 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 6/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 7/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 8/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 9/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000%

Interesting.. I did not have my system with me, so let me get back to you next week. 
In the meantime, could you determine whether the fp-32 baseline under your training configuration would converge? From the log you provided, it seems not enough spikes were delivered to the classification layer.

I will try to reproduce the problem once I get access to run the codes. Please keep me posted on this issue. Thank you.

**Comment by Jealousc11gx** [comment]
> > One more question... I failed to train the vgg9 when I set quantization bit to 8 <img alt="d6a7ec516901a8e87f601f67922d9a8" width="1010" src="https://private-user-images.githubusercontent.com/67589956/345525456-31fc02c9-3f50-4eec-adfc-1272fa18259f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwNjU4NDcsIm5iZiI6MTcyMDA2NTU0NywicGF0aCI6Ii82NzU4OTk1Ni8zNDU1MjU0NTYtMzFmYzAyYzktM2Y1MC00ZWVjLWFkZmMtMTI3MmZhMTgyNTlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA0VDAzNTkwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmMGE2ODhmYzk4MzU3MTkzOGNiMjg2M2VmMGEzM2I3NzMzYTgxMjMxNjg1NDUxN2JkYWQyZjM1Y2UyZTA0MDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uR5qliaX4bghW3DnabbceNQpOQfB8UPUJtqKRRL7k08">
> > root@autodl-container-bb8c4f8e12-9b6fb58b:~/autodl-tmp/quant_snn# python train_snn.py -uq -wq -share -sft_rst --arch vgg9 ********** SNN simulation parameters ********** Namespace(batch_size=256, lr=0.001, arch='vgg9', dataset_dir='./dataset/', dataset='cifar10', optim='adam', leak_mem=0.5, th=0.5, rst='hard', T=4, uq=True, bq=False, wq=True, share=True, sft_rst=True, conv_b=False, bn_a=False, xa=False, ts=False, epoch=150, num_workers=4, train_display_freq=1, test_display_freq=1) Files already downloaded and verified Files already downloaded and verified quant bw for w: 8 quant bw for u: 8 Train Epoch: 0/150 Loss: 2.303124 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 1/150 Loss: 2.302700 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 2/150 Loss: 2.302739 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 3/150 Loss: 2.302704 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 4/150 Loss: 2.302775 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 5/150 Loss: 2.302424 Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 6/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 7/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 8/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000% Train Epoch: 9/150 Loss: nan Accuracy: 10.000% Best Accuracy: 10.000%
> 
> Interesting.. I did not have my system with me, so let me get back to you next week. In the meantime, could you determine whether the fp-32 baseline under your training configuration would converge? From the log you provided, it seems not enough spikes were delivered to the classification layer.
> 
> I will try to reproduce the problem once I get access to run the codes. Please keep me posted on this issue. Thank you.

Thank you for your help! under the training configuration of 2, 3, 4bits and fp32 ,  it would converge. But when I set the bit more than 5, It will not. Here is the acc and loss pic

**Comment by RuokaiYin** [comment]
Hello, 
I just got access to the system. I am working on the issue. It seems like there's a bug in the code that makes the network fail to converge with more than 5-bit quantization. I will get back to you soon.

**Comment by Jealousc11gx** [comment]
Hi, 
I just modified my network structure and run the mint model then store it, however , I find that the weights and membrane potential are full precision. If i want to deploy the mint model in my hardware, how should I store the quantizated weights and membrane potentials and inference with quantizated weight and membrane potential?

**Comment by RuokaiYin** [comment]
Cool, I am happy the training works! If you examine the model directly on PyTorch, the weights and membrane potential should still be at full precision. 
You need to extract the quantized weights from the model manually. I believe there's a w_q_inference function under the training_utils.py. The w_hat it returns is the quantized weights you want.
I assume you are sharing the quantization scale; in that case, you also need to use the beta to scale your firing threshold at deployment time to retain the accuracy.

The membrane potential depends on your deployment strategy. Usually, it's a dynamic on-chip quantity, so you can clamp it on the hardware on the fly during the inference time. But remember to clamp only the residual membrane potential. 

Some of the details can be found in the paper. Sections IV. and V. should give you some insights. One more note here, currently, the deployment should only be performed on VGG9 or any other SNN models that do not have the batchnorm since we have not performed batchnorm folding in this work.

I am very excited that you decided to deploy MINT on the hardware. We also planned to try that but had to steer the wheel due to some other work. If you need further assistance, feel free to email me at ruokai.yin@yale.edu. I am happy to provide you with more assistance from there.

**Comment by RuokaiYin** [comment]
I am closing this issue for now. Feel free to contact me directly if there are any further questions. I have not tested the convergence problem yet; I will open another issue for that until I clear the problem.
